File: rust_exercises/leetcode/Problem749/Solution.java
class Solution {
    public int containVirus(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;

        int countWalls = 0;
        while (true) {
            boolean[][] visited = new boolean[m][n];
            Set<Integer> maxThreat = new HashSet<>();
            Set<Integer> maxBlock = new HashSet<>();
            Set<Integer> toInfect = new HashSet<>();
            int maxWalls = 0;
            
            for (int i = 0; i < m; i += 1) {
                for (int j = 0; j < n; j += 1) {
                    if (grid[i][j] == 1 && !visited[i][j]) {
                        Set<Integer> threat = new HashSet<>();
                        Set<Integer> block = new HashSet<>();
                        int walls = dfs(i, j, grid, visited, threat, block);
                        
                        if (threat.size() > maxThreat.size()) {
                            toInfect.addAll(maxThreat);
                            maxThreat = threat;
                            maxBlock = block;
                            maxWalls = walls;
                        } else {
                            toInfect.addAll(threat);
                        }
                    }
                }
            }
            
            if (maxThreat.isEmpty()) {
                break;
            }
            
            for (int pos: toInfect) {
                grid[pos / n][pos % n] = 1;
            }
            
            for (int pos: maxBlock) {
                grid[pos / n][pos % n] = 2;
            }
            
            countWalls += maxWalls;
        }
        
        return countWalls;
    }
    
    private int dfs(int i, int j, int[][] grid, boolean[][] visited, Set<Integer> threat, Set<Integer> block) {
        int m = grid.length;
        int n = grid[0].length;
        
        visited[i][j] = true;
        block.add(i * n + j);
        
        int[][] moves = new int[][]{{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        int walls = 0;
        
        for (int k = 0; k < 4; k += 1) {
            int ii = i + moves[k][0];
            int jj = j + moves[k][1];
            
            if (valid(ii, jj, m, n)) {
                if (grid[ii][jj] == 0) {
                    walls += 1;
                    threat.add(ii * n + jj);
                } else if (grid[ii][jj] == 1 && !visited[ii][jj]) {
                    walls += dfs(ii, jj, grid, visited, threat, block);
                }
            }
        }
        
        return walls;
    }
    
    
    
    private boolean valid(int i, int j, int m, int n) {
        return i >= 0 && j >= 0 && i < m && j < n;
    }
}

File: rust_exercises/leetcode/Problem749/solution.rs
use std::collections::HashSet;

impl Solution {
    pub fn contain_virus(grid: Vec<Vec<i32>>) -> i32 {
        let m = grid.len();
        let n = grid[0].len();
        let mut grid = grid;
        
        let mut count_walls = 0;
        loop {
            let mut visited = vec![vec![false; n]; m];
            let mut max_threat = HashSet::new();
            let mut max_block = HashSet::new();
            let mut max_walls = 0;
                    
            let mut to_infect = HashSet::new();
            for i in 0..m {
                for j in 0..n {
                    if grid[i][j] == 1 && !visited[i][j] {
                        let mut threat = HashSet::new();
                        let mut block = HashSet::new();
                        let walls = dfs(i, j, &mut grid, &mut visited, &mut threat, &mut block);
                        
                        if threat.len() > max_threat.len() {
                            for x in max_threat.into_iter() {
                                to_infect.insert(x);
                            }
                            max_threat = threat;
                            max_block = block;
                            max_walls = walls;
                        } else {
                            for x in threat.into_iter() {
                                to_infect.insert(x);
                            }
                        }
                    }
                }
            }
            
            if max_threat.is_empty() {
                break;
            }
            for (i, j) in to_infect.into_iter() {
                grid[i][j] = 1;
            }
            for (i, j) in max_block.into_iter() {
                grid[i][j] = 2;
            }
            count_walls += max_walls;
        }
        
        count_walls
    }
}

fn valid(i: i32, j: i32, m: usize, n: usize) -> bool {
    i >= 0 && j >= 0 && (i as usize) < m && (j as usize) < n
}

fn dfs(i: usize, j: usize, grid: &mut Vec<Vec<i32>>, visited: &mut Vec<Vec<bool>>, threat: &mut HashSet<(usize, usize)>, block: &mut HashSet<(usize, usize)>) -> i32 {
    let m = grid.len();
    let n = grid[0].len();
    
    visited[i][j] = true;
    block.insert((i, j));
    
    let moves = vec![(-1, 0), (1, 0), (0, -1), (0, 1)];
    let mut walls = 0;
    
    for (di, dj) in moves.into_iter() {
        let ii = i as i32 + di;
        let jj = j as i32 + dj;
        
        if valid(ii, jj, m, n) {
            let ii = ii as usize;
            let jj = jj as usize;
            if grid[ii][jj] == 0 {
                walls += 1;
                threat.insert((ii, jj));
            } else if grid[ii][jj] == 1 && !visited[ii][jj] {
                walls += dfs(ii, jj, grid, visited, threat, block);
            }
        }
    }
    
    walls
}
